### PR TITLE
[DDW-691] Implement custom browser protocol to launch Daedalus

### PR DESCRIPTION
**Pull request by tomislavhoracek**
This PR introduces custom browser protocol handlers that will launch Daedalus with a specific set of parameters.

## Todos

- [x] Check if launching works on builds
- [x] Carryover data via query parameters and parse
- [x] Trigger an action that goes to a hardcoded route (summary page of the first wallet)
- [x] Check what are all possible parameters Daedalus can receive from an external source using a custom protocol
- [x] Check is protocol naming (`cardano`) good enough -> We will use `web+cardano` as is defined by https://github.com/cardano-foundation/CIPs/blob/master/CIP-0013/CIP-0013.md
- [x] Fix protocol not working on [Windows](https://input-output-rnd.slack.com/archives/C7PJ29FME/p1627328516438700) - @michalrus  
- [x] Fix protocol not working on [Linux](https://input-output-rnd.slack.com/archives/C7PJ29FME/p1627331638443800) - @michalrus 
- [x] Fix initial launch via `open-url` on macOS - @michalrus 
- [ ] Review and discussion related to the research https://docs.google.com/document/d/1ToezIl-kotoVyD3U5_mRav3WTirp3ag7nIBfgHI7ghM/edit?usp=sharing (@dmitrii-gaico @gabriela-ponce @nikolaglumac )

**Note:**
For now (MPV) we are handling basic transactions using a custom protocol. 

```
cardanourn = "web+cardano:" paymentref
paymentref = cardanoaddress [ "?" amountparam ]
cardanoaddress = *(base58 | bech32)
amountparam = "amount=" *digit [ "." *digit ]
```
_Example:_ 
`web+cardano:addr_test1qrakn4h9wkq96rx2eyfhn7mydpf285pg80dma203sr3qmulqqapc66zkutqnm0zk55gvs2drll6nune7eh24ax7ld3yqy53sf0?amount=5.000000`

_Pattern:_ 
web+cardano:`your_address`?amount=5.000000

---

## Test Scenarios

**Scenario 1 - Submit URL to open Daedalus**
```gherkin
Given that Daedalus "Network" is closed 
When I submit a custom protocol URL on "Browser"
Then I can see that the browser tries to run the corresponding instance of Daedalus

| Network   | Browser |
| Mainnet   | Edge    |
| Flight    | Edge    |
| Testnet   | Edge    |
| Mainnet   | Safari  |
| Flight    | Safari  |
| Testnet   | Safari  |
| Mainnet   | Chrome  |
| Flight    | Chrome  |
| Testnet   | Chrome  |
| Mainnet   | Firefox |
| Flight    | Firefox |
| Testnet   | Firefox |
| Mainnet   | Opera   |
| Flight    | Opera   |
| Testnet   | Opera   |

```

**Scenario 2 - Open Daedalus from URL (with at least a wallet)**
```gherkin
Given that Daedalus is closed 
And it is synced
And it has at least a synced wallet
When I submit a custom protocol URL on the browser
And I allow the browser to open Daedalus
Then I can see it opens on the "Send" tab from the first wallet
And the fields are filled according to the data on the URL
```

**Scenario 3 - Open Daedalus from URL (no wallets)**
```gherkin
Given that Daedalus is closed 
And it is synced
And it doesn't have a wallet
When I submit a custom protocol URL on the browser
And I allow the browser to open Daedalus
Then I can see Daedalus opens
And no further action is performed
```

**Scenario 4 - Submit URL to create a transaction on Daedalus (with at least a wallet)**
```gherkin
Given that Daedalus is open 
And it is synced
And it has at least a synced wallet
When I submit a custom protocol URL on the browser
Then I can see that the corresponding instance of Daedalus is displayed on the front 
And it displays the "Send" tab from the last active wallet wallet
And the fields are filled according to the data on the URL
```

**Scenario 5 - Submit URL to create a transaction on Daedalus (with at least a wallet, node syncing)**
```gherkin
Given that Daedalus is open 
And it is not synced
And it has at least a wallet
When I submit a custom protocol URL on the browser
And it displays the "Send" tab from the last active wallet wallet
And no further action is performed
```

**Scenario 6 - Submit URL to create a transaction on Daedalus (wallets syncing)**
```gherkin
Given that Daedalus is open 
And it has at least a wallet
And the wallets are not synced
When I submit a custom protocol URL on the browser
And it displays the "Send" tab from the last active wallet wallet
And no further action is performed
```

**Scenario 7 - Submit URL to create a transaction on Daedalus (no wallets)**
```gherkin
Given that Daedalus is open 
And it is synced
And it doesn't have a wallet
When I submit a custom protocol URL on the browser
Then I can see that the corresponding instance of Daedalus is displayed on the front
And no further action is performed
```

**Scenario 8 - Submit URL to create a transaction on Daedalus (no wallets, node syncing)**
```gherkin
Given that Daedalus is open 
And it is not synced
And it doesn't have a wallet
When I submit a custom protocol URL on the browser
Then I can see that the corresponding instance of Daedalus is displayed on the front
And no further action is performed
```

**Scenario 9 - Check the correct instance of Daedalus opens**
```gherkin
Given that I have multiple networks of Daedalus installed
And all are synced
When I submit a custom protocol URL that contains an address from a version that is installed
And I allow the browser to open Daedalus
Then I can see the corresponding instance of Daedalus opens
```

**Scenario 10 - Check the correct instance of Daedalus shows the transaction**
```gherkin
Given that I have multiple networks of Daedalus installed
And all are synced
And at least two of them are running at the same time
When I submit a custom protocol URL that contains an address from a version that is installed
And I allow the browser to open Daedalus
Then I can see the corresponding instance of Daedalus is displayed on the front
And it displays the "Send" tab from the last active wallet wallet
And the fields are filled according to the data on the URL 
```

---

## Testing Checklist

**_Notes:_** 
- To test custom protocol use this pattern (check example in PR description):
web+cardano:`your_address`?amount=5.000000
- Paste constructed url to the browser and check test cases
- For now (MPV) we are handling basic transactions using a custom protocol.

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)

The result for all test cases should be: 
1. If you have an active wallet then active wallet send form with pre-filled address and amount should be the result (Case when the app is running)
2. If you have restored wallets but you are not on some wallet screen (e.g. you are on the voting screen) then the first wallet from your list will be chosen as an active wallet and you should see send form with pre-filled address and amount as the result (Case when the app is running)
3. If you don't have restored wallets, nothing will happen (Case when the app is running)
4. If Daedalus is closed by executing custom protocol it should open, wait to be connected and loaded and then will follow some case from (1 - 3) depending on the current state 

- [ ] Check cases from 1 - 4 on all platforms
- [ ] Since Yoroi is also using `web+cardano` as custom protocol, check which app will open first by executing custom protocol in your browser, Yoroi extension or Daedalus, and report here
- [ ] Since we can have multiple Daedalus instances (e.g. Daedalus Mainnet, Daedalus Testnet), we need to check opening priorities. Which app instance will be opened if multiple instances are installed on your machine. Check and report here. 

- _Test cases with multiple instances need to be checked because we have few opened proposals related to that. For example, can we pass network parameters to custom protocol parameters or maybe we can say that we want to use daedalus+web instead of web+cardano to avoid conflicts between Daedalus and Yoroi._ We can't change custom protocol URL patterns because they are strictly defined in https://github.com/cardano-foundation/CIPs/blob/master/CIP-0013/CIP-0013.md but we can add some suggestions and concerns :)  Please check this thread: https://input-output-rnd.slack.com/archives/GGKHNC5V4/p1623108628028900

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by SebastienGllmt** [comment]
Have you read https://github.com/cardano-foundation/CIPs/blob/master/CIP-0013/CIP-0013.md ? I noticed this isn't compatible with it

**Comment by tomislavhoracek** [comment]
@SebastienGllmt This is still in progress and for now we just implemented a basic handler to check if it will work once the app is built. We will follow documentation once we are sure that our setup works properly. Thanks!
cc @nikolaglumac

**Comment by dmitrii-gaico** [comment]
@tomislavhoracek @daniloprates Hello. QA team had a quick looks and please see the short report below.

Linux Testnet installed.
Daedalus open\closed. Chrome/Firefox - after hitting link a search is performed.(No Yoroi extension installed) 
Daedalus open\closed. Chrome/Firefox - after hitting link Yoroi is opened.(Yoroi extension installed) 

Windows Testnet and Mainnet installed.
Daedalus open\closed. Chrome/Firefox - after hitting link user suggested to open the Latest installed build. If the latest is Mainnet then it is suggested to open. However after confirming the Daedalus improperly started message is displayed. No Daedalus started.(No Yoroi extension installed) 
Daedalus open\closed. Chrome/Firefox - after hitting link Yoroi is opened.(Yoroi extension installed). 

Strange behavior related to Yoroi Extension :
If you have extension installed and you tried to run web link then it is opened in Yoroi. But once you uninstalled Yoroi extension Chrome is still trying to run the link as Extension but since it is gone Chrome is showing message that link is blocked and you can not see Open Daedalus dialog. In order to fix this you need to re-install Chrome at Windows \ Re-install and delete configs at Linux.

@input-output-hk/daedalus-qa Please add your findings for MacOS and anything related.

Thank you!

**Comment by daniloprates** [comment]
@dmitrii-gaico my first test on testnet worked flawlessly. Do I need to do something else to reproduce the issue?

https://user-images.githubusercontent.com/1504716/127016332-ad91822c-4325-4199-8148-7af95ade0ced.mov

**Comment by daniloprates** [comment]
@dmitrii-gaico if Daedalus is closed, chrome will open it, but it won't go to the send screen

https://user-images.githubusercontent.com/1504716/127017424-a753af13-aa85-4073-ae79-ee577bf3b56c.mov

**Comment by gabriela-ponce** [comment]
Here are my findings on macOS BigSur on build 18656 with different browsers:

**Chrome with Yoroi extension**:
- Opens Daedalus after submitting the address. If Daedalus is open, [it creates the transaction](https://drive.google.com/file/d/1OpC5VUYVABH5FgjGlHu4QpWicL-AfEew/view?usp=sharing). If it's closed, [it only opens Daedalus, doesn't create the transaction](https://drive.google.com/file/d/1H0B1iibX09DiFwBXDV9kIrKwveWP-hLh/view?usp=sharing).

**Firefox without Yoroi extension**:
- Same results as Chrome when Daedalus is [closed](https://drive.google.com/file/d/1m4u6K-v6jqTEdRcpOicBvw5zWJobZoPM/view?usp=sharing) and [open](https://drive.google.com/file/d/18PLut0QxVWwXjCF9uhfbzejqzcv8siXC/view?usp=sharing).

**Safari without Yoroi extension**:
- Always performs a search, it doesn't ask to open Daedalus, no matter if it's [closed](https://drive.google.com/file/d/153eBtxByCC4k7hYsOk2lUPX8wqXCrYKc/view?usp=sharing) or [open](https://drive.google.com/file/d/1QqvBfYxDGkct82c8okeWPmwZ2yVN4iBY/view?usp=sharing).

**Opera without Yoroi extension**:
- Usually performs a search, it doesn't ask to open Daedalus, no matter if [closed](https://drive.google.com/file/d/1FmUQELtwfSa7xO33ZUTG__iYM_EddZuo/view?usp=sharing) or [open](https://drive.google.com/file/d/1FTkR6xHE76Yj82BN0N766xU1u2rE51wZ/view?usp=sharing). Sometimes [it opens Daedalus and created the transaction](https://drive.google.com/file/d/1rOm8UU9TGwqIcjHCNYJ7Ksb12bxxPHsz/view?usp=sharing), but not sure how to trigger this, need to investigate the configuration.

Will update with more information about other scenarios later.

**Comment by daniloprates** [comment]
> @tomislavhoracek @daniloprates Hello. QA team had a quick look and please see the short report below.
> 
> **Linux Testnet installed.**
> Daedalus open\closed. Chrome/Firefox - after hitting link a search is performed.(No Yoroi extension installed)
> Daedalus open\closed. Chrome/Firefox - after hitting link Yoroi is opened.(Yoroi extension installed)
> 
> **Windows Testnet and Mainnet installed.**
> Daedalus open\closed. Chrome/Firefox - after hitting link user suggested to open the Latest installed build. If the latest is Mainnet then it is suggested to open. However after confirming the Daedalus improperly started message is displayed. No Daedalus started.(No Yoroi extension installed)
> Daedalus open\closed. Chrome/Firefox - after hitting link Yoroi is opened.(Yoroi extension installed).
> 
> **Strange behavior related to Yoroi Extension :**
> If you have extension installed and you tried to run web link then it is opened in Yoroi. But once you uninstalled Yoroi extension Chrome is still trying to run the link as Extension but since it is gone Chrome is showing message that link is blocked and you can not see Open Daedalus dialog. In order to fix this you need to re-install Chrome at Windows \ Re-install and delete configs at Linux.
> 
> @input-output-hk/daedalus-qa Please add your findings for MacOS and anything related.
> 
> Thank you!

@dmitrii-gaico could you paste the errors here?

**Comment by daniloprates** [comment]
> @daniloprates @dmitrii-gaico Here are my findings on build 18656 with different browsers and platforms:
> 
> **macOS**
> 
> _Chrome with Yoroi extension_
> 
> * Opens Daedalus after submitting the address. If Daedalus is open, [it creates the transaction](https://drive.google.com/file/d/1OpC5VUYVABH5FgjGlHu4QpWicL-AfEew/view?usp=sharing). If it's closed, [it only opens Daedalus, doesn't create the transaction](https://drive.google.com/file/d/1H0B1iibX09DiFwBXDV9kIrKwveWP-hLh/view?usp=sharing).
> 
> _Firefox without Yoroi extension_
> 
> * Same results as Chrome when Daedalus is [closed](https://drive.google.com/file/d/1m4u6K-v6jqTEdRcpOicBvw5zWJobZoPM/view?usp=sharing) and [open](https://drive.google.com/file/d/18PLut0QxVWwXjCF9uhfbzejqzcv8siXC/view?usp=sharing).
> 
> _Safari without Yoroi extension_
> 
> * Always performs a search, it doesn't ask to open Daedalus, no matter if it's [closed](https://drive.google.com/file/d/153eBtxByCC4k7hYsOk2lUPX8wqXCrYKc/view?usp=sharing) or [open](https://drive.google.com/file/d/1QqvBfYxDGkct82c8okeWPmwZ2yVN4iBY/view?usp=sharing).
> 
> _Opera without Yoroi extension_
> 
> * Usually performs a search, it doesn't ask to open Daedalus, no matter if [closed](https://drive.google.com/file/d/1FmUQELtwfSa7xO33ZUTG__iYM_EddZuo/view?usp=sharing) or [open](https://drive.google.com/file/d/1FTkR6xHE76Yj82BN0N766xU1u2rE51wZ/view?usp=sharing). Sometimes [it opens Daedalus and created the transaction](https://drive.google.com/file/d/1rOm8UU9TGwqIcjHCNYJ7Ksb12bxxPHsz/view?usp=sharing), but not sure how to trigger this, need to investigate the configuration.
> 
> **Windows**
> 
> _All browsers, no Yoroi extension (Edge/Firefox/Opera/Chrome)_
> 
> * I was able to reproduce @dmitrii-gaico's findings. The browser asks to run Daedalus, but then there is an error system displayed mentioning that Daedalus was improperly started. It doesn't matter if Daedalus is closed or open. Check videos for [Edge](https://drive.google.com/file/d/1I-lTkPK8AUv94FKEpRbZPmOXCG9l7wca/view?usp=sharing), [Firefox](https://drive.google.com/file/d/1Tb-Mypm8BUCAToKkiTrtL6tNDkLc9i6C/view?usp=sharing), [Chrome](https://drive.google.com/file/d/1QFFHBUpJ7_HEVqvv_hrl4-cYpjdSVj_g/view?usp=sharing) and [Opera](https://drive.google.com/file/d/1oemhFxeFopaTiYf9QrpQPergizXVz6aI/view?usp=sharing).
> 
> _Chrome, with Yoroi extension_
> 
> * Maybe this is because Yoroi was installed after Daedalus (needs more investigation), but Yoroi is open instead of Daedalus. Check [video](https://drive.google.com/file/d/1gQKKQQTABXomHT0aGfjl-sWdz95JjG14/view?usp=sharing).
> 
> **Both Platforms**
> 
> _All browsers_
> Similar to what @dmitrii-gaico mentioned, no matter which one is the address on the url (mainnet, testnet), the browser would always try to open the last one that was executed. Check [this video.](https://drive.google.com/file/d/1jDoDDh2ZVfKRCqnQHmah9NoqUogiYry8/view?usp=sharing)

Thanks @gabriela-ponce!

File: source/main/index.js [hunk 1]
```diff
@@ -257,6 +259,37 @@ const onAppReady = async () => {
...
+      argv: process.argv,
+    });
+    app.setAsDefaultProtocolClient('web+cardano', process.execPath, [
+      process.argv[1],
+    ]);
+  } else {
+    logger.info('[Custom-Protocol] Set Mac / Linux protocol params');
+    app.setAsDefaultProtocolClient('web+cardano');
+    childProcess.exec(
+      'xdg-mime default Daedalus*.desktop x-scheme-handler/web+cardano'
```
Comment (cleverca22) [inline source/main/index.js:274]: `xdg-mime default Daedalus*.desktop` must be ran with the right name for the cluster

the install script on linux does that for you, so the JS code doesnt need to run `xdg-mime` on linux

File: source/main/index.ts [hunk 2]
```diff
@@ -256,7 +263,57 @@ const onAppReady = async () => {
     client.create(mainWindow);
   }
 
+  // Register custom browser protocol
+
+  // TODO: this is unethical! user hostility – don’t touch my default programs, esp. when not asking me – @michalrus
```
Comment (michalrus) [inline source/main/index.ts:268]: I think this shouldn’t happen like that.

Right now, we re-associate `web+cardano://` to Daedalus on Windows and macOS on each application start, disregarding what the user might want.

Linux is set up correctly – associating `web+cardano://` once, in the installer.

I think we should do the same thing for Windows and macOS – associate *once* during installation, and never re-associate.

**Review by michalrus — CHANGES_REQUESTED** [review]

File: source/main/index.ts (hunk 2 above)
Comment (michalrus) [reply source/main/index.ts:268]: /cc @angerman – a bit similar to auto-adding the app to macOS dock, which you found hostile – https://input-output-rnd.slack.com/archives/GJMEW8FV3/p1647944259851109?thread_ts=1647861796.476749&cid=GJMEW8FV3

File: source/renderer/app/components/wallet/WalletSendForm.tsx [hunk 3]
```diff
@@ -76,6 +76,8 @@ type Props = {
   walletName: string;
   onTokenPickerDialogOpen: (...args: Array<any>) => any;
   onTokenPickerDialogClose: (...args: Array<any>) => any;
+  customProtocolParameters: any | null | undefined; // TBD once all custom protocol parameters are declared
+  resetCustomProptocolParams: Function | null | undefined;
```
Comment (michalrus) [inline source/renderer/app/components/wallet/WalletSendForm.tsx:80]: typo: pro**p**tocol

File: source/main/index.ts [hunk 4]
```diff
@@ -317,18 +374,93 @@ const onAppReady = async () => {
...
+});
+
 // Make sure this is the only Daedalus instance running per cluster before doing anything else
-const isSingleInstance = app.requestSingleInstanceLock();
+const isFirstInstance = app.requestSingleInstanceLock();
 
-if (!isSingleInstance) {
-  app.quit();
+if (!isFirstInstance) {
+  safeExitWithCode(0);
```
Comment (michalrus) [inline source/main/index.ts:436]: I’m really not sure why this change was introduced – if `safeWithExitCode` for second instance will just `app.exit(0)` anyway… (and had to be modified to do just that, or else there are errors about `log.transports.file.stream` being undefined.

**Review by michalrus — CHANGES_REQUESTED** [review]
9 tests are failing in `yarn test:jest`: 
```
FAIL  source/renderer/app/components/wallet/WalletSendForm.spec.tsx (6.566 s)
```

**Comment by marcin-mazurek** [comment]
Closing as agreed with @tomislavhoracek on Slack.

**Pull request closed without merging**